backend/app/main.py

```
@@ -448,6 +503,13 @@ async def restore_text(request: RestoreRequest):
         
         return RestoreResponse(restored_text=restored_text)
         
+    except ValueError as e:
+        # Handle validation errors (e.g., tokens not found in redacted text)
+        logger.warning(f"Validation error during restoration: {e}")
+        raise HTTPException(
+            status_code=400,
+            detail=f"Invalid restoration request: {str(e)}"
+        )
```

Ruff (0.12.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)